### PR TITLE
4762: Add capabilities selector page

### PR DESCRIPTION
**Pull request by kat1906**

### CODE REVIEW

File: app/pages/browse-solutions/template.njk [hunk 1]
```diff
@@ -20,7 +20,7 @@
     <div class="nhsuk-grid-row nhsuk-promo-group" data-test-id="browse-solutions-promo-group">
       <div class="nhsuk-grid-column-one-half nhsuk-promo-group__item nhsuk-u-padding-left-0" data-test-id="all-solutions-promo">
         {{ promo({
-          "href": "/solutions/all",
+          "href": "/solutions/capability-selector",
```
Comment (alibombaye) [inline app/pages/browse-solutions/template.njk]: I know this may have been partly my fault but can we stick with `capabilities`

Comment (kat1906) [reply app/pages/browse-solutions/template.njk]: if by partly you mean all then yeah 😄

File: app/pages/capability-selector/capabilitySelectorPageContext.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import content from './manifest.json';
+
+export const createCapabilitySelectorPageContext = ({ capabilities }) => (
+  {
+    ...content,
+    capabilities: capabilities.capabilities.reduce((acc, capability, i) => {
```
Comment (alibombaye) [inline app/pages/capability-selector/capabilitySelectorPageContext.js]: I'm not capable of handling so many capabilities in one line. Is there a way to tidy this up a little please?

File: app/pages/capability-selector/capabilitySelectorPageContext.test.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+import { createCapabilitySelectorPageContext } from './capabilitySelectorPageContext';
+import manifest from './manifest.json';
+
+describe('createCapabilitySelectorPageContext', () => {
+  it('should create a context for the capability selector page', () => {
```
Comment (alibombaye) [inline app/pages/capability-selector/capabilitySelectorPageContext.test.js]: Don't think we need this test. All your other test are more specific and cover all this.

File: app/pages/capability-selector/capabilitySelectorPageContext.test.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        name: 'Prescribing2',
+        isFoundation: true,
+      }],
+    };
+    const context = createCapabilitySelectorPageContext({ capabilities });
+
+    expect(context.capabilities.column1.length).toEqual(context.capabilities.column2.length);
+  });
+
+  it('should create a context for the capability-selector page with an extra one in coluumn 1 when an odd number passed in', () => {
```
Comment (alibombaye) [inline app/pages/capability-selector/capabilitySelectorPageContext.test.js]: `coluumn `

File: app/pages/capability-selector/components/macro.njk [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  <div id="capability-selector" data-test-id="capability-selector">
+    <div data-test-id="capability-checkbox-column-1" class="nhsuk-grid-column-one-half">
+      {{ checkboxes({
+        "idPrefix": "column-1",
+        "items": params.capabilities.column1
+      }) }}
+    </div>
+    <div data-test-id="capability-checkbox-column-2" class="nhsuk-grid-column-one-half">
+      {{ checkboxes({
+        "idPrefix": "column-1",
```
Comment (alibombaye) [inline app/pages/capability-selector/components/macro.njk]: same id?

File: app/pages/capability-selector/controller.test.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import { getCapabilitiesContext } from './controller';
+import { ApiProvider } from '../../apiProvider';
+import manifest from './manifest.json';
+
+jest.mock('../../apiProvider');
+
+describe('capability-selector controller', () => {
+  describe('getCapabilitiesContext', () => {
+    it('should return the context when preview data is returned by the ApiProvider', async () => {
```
Comment (alibombaye) [inline app/pages/capability-selector/controller.test.js]: preview data?

File: app/pages/capability-selector/template.njk [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    </div>
+    <h1 class="nhsuk-u-margin-top-6 nhsuk-u-margin-bottom-4" data-test-id="capability-selector-page-title">{{ title }}</h1>
+    <div class="nhsuk-grid-row">
+      <h2 class="nhsuk-u-font-size-22 nhsuk-u-padding-bottom-5 nhsuk-grid-column-two-thirds" data-test-id="capability-selector-page-description">{{ description }}</h2>
+      {{ capabilitySelector({
+        capabilities: capabilities
+      }) }}
+    </div>
+    <div data-test-id="capability-selector-continue-button">
+      {# TODO: change href of button #}
```
Comment (alibombaye) [inline app/pages/capability-selector/template.njk]: we going to change this in the next ticket? when we do the post?

File: app/pages/capability-selector/template.test.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    path: 'pages/capability-selector/template.njk',
+  },
+};
+
+describe('capability-selector page', () => {
+  it('should render a backLink to the solutions page', createTestHarness(setup, (harness) => {
+    const context = {};
+
+    harness.request(context, ($) => {
+      const homepageBackLink = $('[data-test-id="go-back-link"]');
```
Comment (alibombaye) [inline app/pages/capability-selector/template.test.js]: just `backLink`?

File: app/pages/capability-selector/template.test.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      expect(homepageBackLink.length).toEqual(1);
+      expect(homepageBackLink.text().trim()).toEqual('Go back to previous page');
+      expect($(homepageBackLink).find('a').attr('href')).toEqual('/solutions');
```
Comment (alibombaye) [inline app/pages/capability-selector/template.test.js]: why the `$` here?

File: app/pages/capability-selector/ui.test.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  await pageSetup(t);
+
+  const capabilitySelectorComponent = Selector('[data-test-id="capability-selector"]');
+  const column1 = capabilitySelectorComponent.find('[data-test-id="capability-checkbox-column-1"]');
+  const column2 = capabilitySelectorComponent.find('[data-test-id="capability-checkbox-column-2"]');
+
+  await t
+    .expect(capabilitySelectorComponent.exists).ok()
+    .expect(column1.exists).ok()
+    .expect(column1.find('label').count).eql(2)
```
Comment (alibombaye) [inline app/pages/capability-selector/ui.test.js]: why not use `.length` instead :trollface:

File: app/routes.test.js [hunk 11]
```diff
@@ -29,169 +29,185 @@ const mockGetPublicSolutionById = {
...
-    const app = new App().createApp();
-    app.use('/', routes);
-
-    return request(app)
-      .get('/healthcheck')
-      .expect(200)
-      .then((res) => {
-        expect(res.text).toBe('Public browse is running!!!');
-      });
+describe('routes', () => {
```
Comment (alibombaye) [inline app/routes.test.js]: why this?

File: app/pages/capability-selector/capabilitySelectorPageContext.test.js (hunk 3 above)
Comment (kat1906) [reply app/pages/capability-selector/capabilitySelectorPageContext.test.js]: are you suggesting we delete a test?? 😱

Comment (alibombaye) [reply app/pages/capability-selector/capabilitySelectorPageContext.test.js]: I know. Shock right!!! 🙀

File: app/pages/capability-selector/template.njk (hunk 7 above)
Comment (kat1906) [reply app/pages/capability-selector/template.njk]: yep I thought so or we could do it in it's own ticket

File: app/pages/capability-selector/ui.test.js (hunk 10 above)
Comment (kat1906) [reply app/pages/capability-selector/ui.test.js]: why didn't I think of that!

File: app/pages/capability-selector/template.test.js (hunk 9 above)
Comment (kat1906) [reply app/pages/capability-selector/template.test.js]: no idea - I am surprised it still works

File: app/routes.test.js (hunk 11 above)
Comment (kat1906) [reply app/routes.test.js]: so they are together?

File: app/pages/capability-selector/capabilitySelectorPageContext.test.js (hunk 3 above)
Comment (kat1906) [reply app/pages/capability-selector/capabilitySelectorPageContext.test.js]: the whole file?

File: app/pages/capability-selector/capabilitySelectorPageContext.js (hunk 2 above)
Comment (kat1906) [reply app/pages/capability-selector/capabilitySelectorPageContext.js]: of course

File: app/routes.test.js (hunk 11 above)
Comment (alibombaye) [reply app/routes.test.js]: oh yeah 😂

File: app/pages/capability-selector/capabilitySelectorPageContext.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import content from './manifest.json';
```
Comment (alibombaye) [inline app/pages/capability-selector/capabilitySelectorPageContext.js]: 😨 should we change the name of these files to `capabilities` too?

Comment (kat1906) [reply app/pages/capability-selector/capabilitySelectorPageContext.js]: the reason I didn't is because it isn't actually a capabilities page

Comment (alibombaye) [reply app/pages/capability-selector/capabilitySelectorPageContext.js]: I meant as in `capabilitiesSelectorPageContext`

**Pull request merged**